node/lifecycle_test.go

```diff
@@ -0,0 +1,211 @@
+	"k8s.io/apimachinery/pkg/watch"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes/fake"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/record"
+	watchutils "k8s.io/client-go/tools/watch"
+	"k8s.io/klog"
+)
+
+var (
+	_ record.EventRecorder = (*fakeDiscardingRecorder)(nil)
+)
+
+const (
+	// There might be a constant we can already leverage here
+	testNamespace        = "default"
+	informerResyncPeriod = time.Duration(1 * time.Second)
+	testNodeName         = "testnode"
+	podSyncWorkers       = 3
+)
+
+func init() {
+	klog.InitFlags(nil)
+}
+
+// fakeDiscardingRecorder discards all events. Silently.
+type fakeDiscardingRecorder struct {
+}
+
+func (*fakeDiscardingRecorder) Event(object runtime.Object, eventtype, reason, message string) {
+}
+
+func (*fakeDiscardingRecorder) Eventf(object runtime.Object, eventtype, reason, messageFmt string, args ...interface{}) {
+}
+
+func (*fakeDiscardingRecorder) PastEventf(object runtime.Object, timestamp metav1.Time, eventtype, reason, messageFmt string, args ...interface{}) {
+}
+
+func (*fakeDiscardingRecorder) AnnotatedEventf(object runtime.Object, annotations map[string]string, eventtype, reason, messageFmt string, args ...interface{}) {
+}
+
+func TestPodLifecycle(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	newLogger := logruslogger.FromLogrus(logrus.NewEntry(logrus.StandardLogger()))
+	logrus.SetLevel(logrus.DebugLevel)
+
+	// Right now, new loggers that are created from spans are broken since log.L isn't set.
+	ctx = log.WithLogger(ctx, newLogger)
+
+	// Create the fake client.
+	client := fake.NewSimpleClientset()
+
+	// This is largely copy and pasted code from the root command
+	podInformerFactory := kubeinformers.NewSharedInformerFactoryWithOptions(
+		client,
+		informerResyncPeriod,
+		kubeinformers.WithNamespace(testNamespace),
+	)
+	podInformer := podInformerFactory.Core().V1().Pods()
+
+	scmInformerFactory := kubeinformers.NewSharedInformerFactory(client, informerResyncPeriod)
+
+	eb := record.NewBroadcaster()
+	eb.StartLogging(log.G(ctx).Infof)
+	eb.StartRecordingToSink(&corev1client.EventSinkImpl{Interface: client.CoreV1().Events(testNamespace)})
+	fakeRecorder := &fakeDiscardingRecorder{}
+
+	secretInformer := scmInformerFactory.Core().V1().Secrets()
+	configMapInformer := scmInformerFactory.Core().V1().ConfigMaps()
+	serviceInformer := scmInformerFactory.Core().V1().Services()
+
+	mockProvider, err := mock.NewMockProviderMockConfig(mock.MockConfig{}, testNodeName, "linux", "1.2.3.4", 0)
+	assert.NilError(t, err)
+	config := PodControllerConfig{
+		PodClient:         client.CoreV1(),
+		PodInformer:       podInformer,
+		EventRecorder:     fakeRecorder,
+		Provider:          mockProvider,
+		ConfigMapInformer: configMapInformer,
+		SecretInformer:    secretInformer,
+		ServiceInformer:   serviceInformer,
+	}
+
+	pc, err := NewPodController(config)
```

Comment:
From about here up, we're mostly setting up.  It would be nice factor out of the test a lot of the setup.

Reply:
That's the idea once we get enough feedback about the test. I plan on adding more tests with this approach shortly.